### PR TITLE
[Azure][Cluster] Only set evictionPolicy to Delete when spot is used

### LINKED ISSUE #46198: [Ray-Clusters] Azure Cluster Autoscaler failing to start worker nodes when using non-spot instances.
### What happened + What you expected to happen

When spinning up workers in a ray cluster using the autoscaler the creation will fail and it will keep trying to initiate new virtual machines and in the process spin up public ip's and network interfaces until the limit is reached.

In the last version the following example with neither priority nor billing profile option set works as intended.

```
# optionally set priority to use Spot instances
#priority: Spot
# set a maximum price for spot instances if desired
#billingProfile:
#    maxPrice: -1
```

When running with the new latest version, I run into the following Issue:
(This is a log of the monitor.log from the headnode)

```
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/node_launcher.py", line 113, in _launch_node
    created_nodes = self.provider.create_node_with_resources_and_labels(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/node_provider.py", line 152, in create_node_with_resources_and_labels
    return self.create_node(node_config, tags, count)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/_azure/node_provider.py", line 247, in create_node
    self._create_node(node_config, tags, count)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/_azure/node_provider.py", line 300, in _create_node
    create_or_update(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 94, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/azure/core/polling/_poller.py", line 270, in wait
    raise self._exception  # type: ignore
  File "/home/ray/anaconda3/lib/python3.9/site-packages/azure/core/polling/_poller.py", line 185, in _start
    self._polling_method.run()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/azure/core/polling/base_polling.py", line 772, in run
    raise HttpResponseError(response=self._pipeline_response.http_response, error=err) from err
azure.core.exceptions.HttpResponseError: (DeploymentFailed) At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.
Code: DeploymentFailed
Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.
Exception Details:	(BadRequest) {
	  "error": {
	    "code": (BadRequest) {
	  "error": {
	    "code": "InvalidParameter",
	    "message": "Eviction policy can be set only on Azure Spot Virtual Machines. For more information, see http://aka.ms/AzureSpot/errormessages.",
	    "target": "billingProfile"
	  }
	}
```

To clarify: This happens when the head node tries to spin up worker nodes.

The current workaround for me is just to use the older version for the head image 
Instead of:
head_image: "rayproject/ray-ml:latest-cpu"
Use:
head_image: "rayproject/ray-ml:2.24.0-cpu'

### Versions / Dependencies

The new Version of the rayproject/ray-ml - latest-cpu. I assume it was build with the new version of 2.30 of ray.

### Reproduction script

I used the example azure-full script but disabled the priority option to NOT request spot instances. (I have not tested if it still crashes if you request spot instances)
Simply run the start command: `ray up -y azure-full.yaml` And wait for the autoscaler to attempt to start a VM.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by kekulai-fredchang**
previous update assumed evictionPolicy on all nodes as 'Delete', which yields error for those azure users that do not launch a Spot instance.  

@gramhagen , @eisber , @ijrsvt 

## Why are these changes needed?

Since Spot instance availability is low in azure, many users will launch with regular instance and encounter error.  

added conditional logic in master azure json file to check if priority is indeed Spot , if so, then yield evictionPolicy as 'Delete', otherwise yield json('null') to skip this policy.

## Related issue number
fixes #46198

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gramhagen** [comment]
This makes sense. Although if the user provides this parameter, I would default to use that first before filling in a default. I'm assuming it's null if not provided? If so you should be able to check if the policy is null and the priority is for spot then use delete otherwise keep the parameter as is.

**Comment by frdchang** [comment]
> This makes sense. Although if the user provides this parameter, I would default to use that first before filling in a default. I'm assuming it's null if not provided? If so you should be able to check if the policy is null and the priority is for spot then use delete otherwise keep the parameter as is.

@gramhagen  new logic was added to check if the user supplied an eviction policy & a spot instance is used, otherwise if spot instance is regular, it passes null for eviction policy.

File: python/ray/autoscaler/_private/_azure/azure-vm-template.json [hunk 1]
```diff
@@ -253,7 +253,7 @@
                     }
                 },
                 "priority": "[parameters('priority')]",
-                "evictionPolicy": "[parameters('evictionPolicy')]",
+                "evictionPolicy": "[if(and(equals(parameters('priority'), 'Spot'), or(equals(parameters('evictionPolicy'), ''), equals(parameters('evictionPolicy'), json('null')))), 'Delete', parameters('evictionPolicy'))]",
```
Comment (gramhagen) [inline python/ray/autoscaler/_private/_azure/azure-vm-template.json:256]:
```suggestion
                "evictionPolicy": "[if(and(equals(parameters('priority'), 'Spot'), equals(parameters('evictionPolicy'), 'Delete')), json('null'), parameters('evictionPolicy'))]",
```

**Review by gramhagen — COMMENTED** [review]
does this work? my eyes are getting crossed looking at the parentheses in the if condition 😵‍💫

**Comment by anyscalesam** [comment]
@kekulai-fredchang  can you address comments here?

**Comment by frdchang** [comment]
> @kekulai-fredchang can you address comments here?

I looked at @gramhagen suggestions, and from my limited azure knowledge it did not seem correct because it will pass json('null') even when 'Delete' is desired.  

I redid the logic as follows, which allows both spot instances with delete default and user defined pass through, and regular instances with evictionPolicy set to json('null').


```plaintext
IF priority IS 'Spot' THEN
    IF evictionPolicy IS empty OR null THEN
        SET evictionPolicy TO 'Delete'
    ELSE
        USE user-defined evictionPolicy
    ENDIF
ELSE
    USE user-defined evictionPolicy
ENDIF

**Review by gramhagen — APPROVED** [review]

**Comment by mjd3** [comment]
I was running into this same issue, but I think this can be a lot simpler. The `evictionPolicy` parameter already defaults to `Delete` in the template, so you can just set it to an empty string in the case of `Regular` priority and use the user-defined `evictionPolicy` if it is provided. I just replaced [this line](https://github.com/ray-project/ray/blob/25dfc3c08a94a9d9769319f30b12d4136879d4cb/python/ray/autoscaler/_private/_azure/azure-vm-template.json#L256) with:
```
"evictionPolicy": "[if(equals(parameters('priority'), 'Spot'), parameters('evictionPolicy'), '')]",
```

**Comment by gramhagen** [comment]
@mjd3 that does seem like the easiest option!

**Comment by mjd3** [comment]
Also, as a workaround until this PR is merged in, you can add `evictionPolicy: null` in your cluster cfg yaml file for any nodes with `priority: Regular`.

**Comment by mjd3** [comment]
@kekulai-fredchang do you agree with [the suggestion](https://github.com/ray-project/ray/pull/46959#issuecomment-2303346034) above? From looking back, it seems like my suggestion mirrors @gramhagen's initial suggestion too. If so, can you update this PR? Would love to get this one merged in.

**Comment by mjd3** [comment]
Thanks @kekulai-fredchang! @anyscalesam this should be ready to merge.

**Comment by anyscalesam** [comment]
@hongchaodeng can you please take a look?

**Comment by bthananjeyan** [comment]
Sorry for introducing this issue in the earlier PR. I confirmed that this fix works for me locally. @ericl @architkulkarni @hongchaodeng would you be able to review this PR?

**Review by kevin85421 — APPROVED** [review]
I don't have an Azure account to test this PR, but some folks have already verified it. Hence, I approve this PR.

**Comment by bthananjeyan** [comment]
@kekulai-fredchang It looks like you could run these instructions to pass the DCO check: [https://github.com/ray-project/ray/pull/46959/checks?check_run_id=30602666160](https://github.com/ray-project/ray/pull/46959/checks?check_run_id=30602666160)

Thanks for fixing this issue. The current temporary workaround I am using for a non-spot head node and spot worker nodes is to have this fix in my local version of ray, which is required to initially start the head node/cluster. The nodes on the cluster can still have the newer ray releases (I am using 2.35 on the cluster).

**Comment by kevin85421** [comment]
> It looks like you could run these instructions to pass the DCO check

The DCO check failing is fine with me. We just need to make sure the premerge check passes.

**Review by jjyao — APPROVED** [review]

**Pull request merged**